### DISCUSSION TITLE
I don't understand the following code, please help me

### DISCUSSION THREAD

**Discussion by piaoyidage** [axios/axios#5555] [Q&A] [answered]
In lib/cancel/CancelToken，when I  debugger cancel request, code not enter

```js
// eslint-disable-next-line func-names
    this.promise.then = onfulfilled => {
      let _resolve;
      // eslint-disable-next-line func-names
      const promise = new Promise(resolve => {
        token.subscribe(resolve);
        _resolve = resolve;
      }).then(onfulfilled);

      promise.cancel = function reject() {
        token.unsubscribe(_resolve);
      };

      return promise;
    };
```

**Comment by DigitalBrainJS** [comment, marked as answer]
This method fixes the native `then` method of CancelToken to avoid a memory leak since the native Promise has no way to cancel the subscription added by the `then` chain. This is very important when you are using a persistent token for multiple requests.
This method is not used internally by Axios. It exists for backward compatibility with the CancelToken specification, where CancelToken extends Promise, so it must have a public `then` method. Axios uses proprietary `subscribe` and `unsubscribe` methods to manage subscriptions on CancelToken. They are not part of the spec, but this is a more resource-efficient way of doing that.

**Reply by piaoyidage** [reply]
If execute the following code, will execute `token.subscribe(resolove)`
```js
source.token.promise.then(_ => console.log('hello axios'));
```
when execute `source.cancel('cancel request')`, will execute 

```js
this.promise.then(cancel => {
      if (!token._listeners) return;

      let i = token._listeners.length;

      while (i-- > 0) {
        token._listeners[i](cancel);
      }
      token._listeners = null;
    });
```

I fell confused, when execute function `promise.cancel`

**Reply by piaoyidage** [reply]
```js
 source.token.promise.then(_ => console.log('hello axios')).cancel(() => console.log('cancel'))
```

Oh, I understand
